service/history/failoverCheck.go

```diff
@@ -61,3 +64,77 @@ func verifyTimerTaskVersion(shard ShardContext, domainID string, version int64,
 	}
 	return true, nil
 }
+
+// load mutable state, if mutable state's next event ID <= task ID, will attempt to refresh
+// if still mutable state's next event ID <= task ID, will return nil, nil
+func loadMutableStateForTransferTask(context *workflowExecutionContext, transferTask *persistence.TransferTaskInfo, metricsClient metrics.Client, logger bark.Logger) (*mutableStateBuilder, error) {
```

Comment:
Why only transfer task?  Won't TimerTask have a similar issue?

Reply:
Is there a way to generalize the check for both timer and transfer tasks?  Can we instead rely on version on the task and version on the mutable state?

Reply:
the main difference is here
```
isDecisionRetry := transferTask.TaskType == persistence.TransferTaskTypeDecisionTask &&
```
vs 
```
isDecisionRetry := timerTask.TaskType == persistence.TaskTypeDecisionTimeout &&
```